secret-engine/src/main/java/com/sequenceiq/cloudbreak/service/secret/service/rotation/VaultRotationExecutor.java

```diff
@@ -2,25 +2,34 @@
 
 import javax.inject.Inject;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.stereotype.Component;
 
 import com.sequenceiq.cloudbreak.rotation.secret.RotationExecutor;
 import com.sequenceiq.cloudbreak.rotation.secret.SecretLocationType;
 import com.sequenceiq.cloudbreak.rotation.secret.SecretRotationException;
+import com.sequenceiq.cloudbreak.service.secret.domain.RotationSecret;
 import com.sequenceiq.cloudbreak.service.secret.service.SecretService;
 import com.sequenceiq.cloudbreak.service.secret.service.rotation.context.VaultRotationContext;
 
 @Component
 public class VaultRotationExecutor implements RotationExecutor<VaultRotationContext> {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(VaultRotationExecutor.class);
+
     @Inject
     private SecretService secretService;
 
     @Override
     public void rotate(VaultRotationContext rotationContext) {
-        rotationContext.getNewSecretMap().forEach((vaultSecret, newSecretValue) -> {
+        rotationContext.getSecretUpdateSupplierMap().forEach((vaultSecret, secretGeneratorClass) -> {
             try {
-                secretService.update(vaultSecret.getSecret(), newSecretValue);
+                if (secretService.getRotation(vaultSecret) == null) {
```

Comment:
I think the RotationSecret won't be null at all only the backup field will be null if a rotation is happening.

I needed to implement this for my change as well. Please take a look the below implementation does what it need to do.

```java

    private static final Logger LOGGER = LoggerFactory.getLogger(VaultRotationExecutor.class);

    @Inject
    private SecretService secretService;

    @Inject
    private Optional<List<SecretGenerator>> secretGenerators;

    private final Map<Class<? extends SecretGenerator>, SecretGenerator> secretGeneratorMap = new ConcurrentHashMap<>();

    @PostConstruct
    public void setUp() {
        secretGenerators.ifPresent(generators -> generators.forEach(secretGenerator ->
                secretGeneratorMap.put(secretGenerator.getClass(), secretGenerator)));
    }

    @Override
    public void rotate(VaultRotationContext rotationContext) {
        rotationContext.getNewSecretMap().forEach((vaultSecret, generator) -> {
            try {
                if (!secretService.getRotation(vaultSecret).isRotation()) {
                    secretService.putRotation(vaultSecret, secretGeneratorMap.get(generator).generate());
                }
            } catch (Exception e) {
                LOGGER.error("Error during {} secret rotation.", vaultSecret, e);
                throw new SecretRotationException(e, getType());
            }
        });
    }

    @Override
    public void rollback(VaultRotationContext rotationContext) {
        rotationContext.getNewSecretMap().forEach((vaultSecret, generator) -> {
            try {
                RotationSecret rotationSecret = secretService.getRotation(vaultSecret);
                if (rotationSecret.isRotation()) {
                    secretService.update(vaultSecret, rotationSecret.getBackupSecret());
                }
            } catch (Exception e) {
                LOGGER.error("Error during {} secret rollback.", vaultSecret, e);
                throw new SecretRotationException(e, getType());
            }
        });
    }

    @Override
    public void finalize(VaultRotationContext rotationContext) {
        rotationContext.getNewSecretMap().forEach((vaultSecret, generator) -> {
            try {
                RotationSecret rotationSecret = secretService.getRotation(vaultSecret);
                if (rotationSecret.isRotation()) {
                    secretService.update(vaultSecret, rotationSecret.getSecret());
                }
            } catch (Exception e) {
                LOGGER.error("Error during {} secret finalization.", vaultSecret, e);
                throw new SecretRotationException(e, getType());
            }
        });
    }

```

Reply:
like it, will rewrite mine